### ISSUE TITLE
Add a field for selecting review action reasons

### ISSUE THREAD

**Issue by bobsilverberg** [mozilla/addons-server#18233]
This is a spin-off from https://github.com/mozilla/addons-server/issues/18168.

When a reviewer is performing certain actions (e.g., rejecting a version, rejecting multiple versions, replying to a review), they should be able to select one or more reasons for the action. This will be enabled via a simple multi-select for the MVP.

This behaviour will generally be present on screens which currently allow for selection of a canned response, so it might make sense to add the multi-select for the action reason to the same template as the canned response select, which currently exists at https://github.com/mozilla/addons-server/blob/master/src/olympia/reviewers/templates/reviewers/review.html#L119.

A couple of questions (for @kewisch / @diox):
1. Can we confirm precisely on which screens this select should be available?
2. Should there be a maximum number of reasons that can be selected? Would at least one reason be required to perform an action, or is it considered optional?
3. Does this need to be behind a flag while we are developing it, or does the fact that it is only visible to reviewers mean that we can implement it in pieces and just tell them it's not ready yet?

**Comment by bobsilverberg** [comment]
One more question for @kewisch, do you envision this as a multi-select input, or as multiple check boxes?

**Comment by tublitzed** [comment]
@bobsilverberg:

2. No maximum. This should be a required field.

re:multi-select or multiple check boxes - whatever is simplest on your end, given that the end result is the same.

**Comment by bobsilverberg** [comment]
> @bobsilverberg:
> 
>     1. No maximum. This should be a required field.
> 
> 
> re:multi-select or multiple check boxes - whatever is simplest on your end, given that the end result is the same.

I think they are pretty much equivalent in terms of simplicity, so it comes down to what's the best UX for the reviewers. Using check boxes would mean all options would be visible at all times, but if there are a lot of options it might be ugly. Using a multi-select I would imagine that one would have to scroll through the list to see all of the options, so that's a tradeoff with that.

Perhaps @kewisch has a particular preference, or maybe @wagnerand has an opinion?

**Comment by kewisch** [comment]
> This is a spin-off from #18168.
> 
> When a reviewer is performing certain actions (e.g., rejecting a version, rejecting multiple versions, replying to a review), they should be able to select one or more reasons for the action. This will be enabled via a simple multi-select for the MVP.
> 
> This behaviour will generally be present on screens which currently allow for selection of a canned response, so it might make sense to add the multi-select for the action reason to the same template as the canned response select, which currently exists at https://github.com/mozilla/addons-server/blob/master/src/olympia/reviewers/templates/reviewers/review.html#L119.
> 
> A couple of questions (for @kewisch / @diox):
> 
>     1. Can we confirm precisely on which screens this select should be available?
This will only need to be on https://reviewers.addons.mozilla.org/en-US/reviewers/review/<guidhere>

> 
>     2. Should there be a maximum number of reasons that can be selected? Would at least one reason be required to perform an action, or is it considered optional?
No maximum as Rachel noted. If we make this a required field, then we'll need to have an "Other" category, as sometimes we'd encounter issues that cannot be specifically mapped to the categories we have. I think that is reasonable though, this would also allow us to figure out if we need to expand the categories.

>     3. Does this need to be behind a flag while we are developing it, or does the fact that it is only visible to reviewers mean that we can implement it in pieces and just tell them it's not ready yet?
Given we have volunteer reviewers we can't easily control I think we'd need to gate this somehow. It would be fine to gate this on just staff/admin reviewer permissions though, we can tell them to leave it alone.

> One more question for @kewisch, do you envision this as a multi-select input, or as multiple check boxes?
Based on the list we have in internal docs, there would be quite a few categories. On the other hand we want this to be easy to click on. I think a `<select multiple>` would work: It isn't a dropdown, you can see multiple options at once, so it is easy to select more than one. Here is a quick and dirty mockup of what I'm thinking, happy to take other ideas into account:

![image](https://user-images.githubusercontent.com/607198/139880167-75dfd69f-1cda-41be-bec9-e1ed8c3af1c4.png)

**Comment by bobsilverberg** [comment]
> > This is a spin-off from #18168.
> 
> > A couple of questions (for @kewisch / @diox):
> > 
> > 1. Can we confirm precisely on which screens this select should be available?
> >
> 
> This will only need to be on [reviewers.addons.mozilla.org/en-US/reviewers/review](https://reviewers.addons.mozilla.org/en-US/reviewers/review/) as far as I am aware. I'm not sure where else it is used. Grouping it with the canned response makes sense to me.
> 

I shouldn't have used the term "screens". What I am really after is "for which actions does this reasons select apply"? What is the complete list of actions that would require that a reason be specified? Looking at the code, it looks like the following actions require comments:
- `public`
- `reject`
- `reject_multiple_versions`
- `reply`
- `super`
- `comment`

I am guessing that maybe not all of those would also *require* a reason?

> > 
> > 3. Does this need to be behind a flag while we are developing it, or does the fact that it is only visible to reviewers mean that we can implement it in pieces and just tell them it's not ready yet?
> > 
> 
> Given we have volunteer reviewers we can't easily control I think we'd need to gate this somehow. It would be fine to gate this on just staff/admin reviewer permissions though, we can tell them to leave it alone.
> 

If we make it so that it isn't required, would it be okay to show the new field to all users, and send some sort of communication to let them know it needn't be used yet? If they do use it, the data would not be stored, but no errors should occur. Does that seem acceptable, or do we need to hide it from all non-staff/admin reviewers?

> > One more question for @kewisch, do you envision this as a multi-select input, or as multiple check boxes?
> 
> Based on the list we have in internal docs, there would be quite a few categories. On the other hand we want this to be easy to click on. I think a `<select multiple>` would work: It isn't a dropdown, you can see multiple options at once, so it is easy to select more than one. Here is a quick and dirty mockup of what I'm thinking, happy to take other ideas into account:
> 
> ![image](https://user-images.githubusercontent.com/607198/139880167-75dfd69f-1cda-41be-bec9-e1ed8c3af1c4.png)

That looks good, thanks!

**Comment by kewisch** [comment]
> I shouldn't have used the term "screens". What I am really after is "for which actions does this reasons select apply"? What is the complete list of actions that would require that a reason be specified? Looking at the code, it looks like the following actions require comments:
> 
>     * `public`
> 
>     * `reject`
> 
>     * `reject_multiple_versions`
> 
>     * `reply`
These actions should have the reasons

> 
>     * `super`
> 
>     * `comment`
These do not require reasons


> If we make it so that it isn't required, would it be okay to show the new field to all users, and send some sort of communication to let them know it needn't be used yet? If they do use it, the data would not be stored, but no errors should occur. Does that seem acceptable, or do we need to hide it from all non-staff/admin reviewers?

That would be fine as well. Thinking about this more, we can also reach out to the volunteers and let them know they shouldn't use it yet, and/or it won't have any effect. As long as someone accidentally using it won't cause something to break, incrementally adding functionality is fine. Therefore I don't think we need any special gating.

**Comment by bobsilverberg** [comment]
> > I shouldn't have used the term "screens". What I am really after is "for which actions does this reasons select apply"? What is the complete list of actions that would require that a reason be specified? Looking at the code, it looks like the following actions require comments:
> > ```
> > * `public`
> > 
> > * `reject`
> > 
> > * `reject_multiple_versions`
> > 
> > * `reply`
> > ```
> 
> These actions should have the reasons
> 
> > ```
> > * `super`
> > 
> > * `comment`
> > ```
> 
> These do not require reasons
> 

Thank you for confirming those.

> > If we make it so that it isn't required, would it be okay to show the new field to all users, and send some sort of communication to let them know it needn't be used yet? If they do use it, the data would not be stored, but no errors should occur. Does that seem acceptable, or do we need to hide it from all non-staff/admin reviewers?
> 
> That would be fine as well. Thinking about this more, we can also reach out to the volunteers and let them know they shouldn't use it yet, and/or it won't have any effect. As long as someone accidentally using it won't cause something to break, incrementally adding functionality is fine. Therefore I don't think we need any special gating.

It sounds like someone needs to draft a communication to go out to reviewers (both volunteers and staff) to explain this new feature, and to also explain that it won't be ready for use immediately, and that we will inform them when it is ready for use. Do you know who might do that (with Caitlin away)?

**Comment by bobsilverberg** [comment]
@kewisch Just bumping my question from above:

"It sounds like someone needs to draft a communication to go out to reviewers (both volunteers and staff) to explain this new feature, and to also explain that it won't be ready for use immediately, and that we will inform them when it is ready for use. Do you know who might do that (with Caitlin away)?"

**Comment by kewisch** [comment]
@bobsilverberg Either Andreas or I would be doing that. Should we send out a message right away?

**Comment by bobsilverberg** [comment]
> @bobsilverberg Either Andreas or I would be doing that. Should we send out a message right away?

@kewisch My plan is to land code that will add the reason select this week, so yes, I think some communication about this would be good prior to that. We can always put off populating the reasons on prod, which will make the select appear empty, and therefore impossible for anyone to select anything anyway.

**Comment by ioanarusiczki** [comment]
@bobsilverberg 

I verified this on AMO dev (as you said on AMO stage the select box is displays but the actions won't require it):
- looking over the queues with an admin account -  Content review queue has "Approve content" button as a first reviewer action displayed but without the reasons box. I don't know if that's intended. 
- for approve, reject, reply, reject multiple versions the reasons are displayed in the rest of the queues and saving the reviewer action will also require a reason. Also all or multiple reasons can be selected. 
- I also verified this with an addon reviewer and content reviewer account and the reasons are displayed in the queues they can access. 
- I don't think reasons will get too long (like the second one) 
![reasons](https://user-images.githubusercontent.com/33448286/141137905-c0bc262e-1798-45cd-a2c5-515f1bab6e31.png)
